### PR TITLE
MCO-371: metrics: aggregate os_image_url_override metric to avoid unbounded cardinality

### PR DESCRIPTION
**Pull request by sinnykumari**
For context, see https://github.com/openshift/cluster-monitoring-operator/pull/1784

### CODE REVIEW

**Comment by sinnykumari** [comment]
From a local cluster, aggregated data looks fine when both master and worker pool has overridedn osImageURL. See [screenshot](https://github.com/sinnykumari/misc-data/blob/main/aggregated_data.png)

@cgwalters PTAL, does this look good to go

**Comment by cgwalters** [comment]
I'll be honest, I don't know the Prometheus stuff very well.  Something that wasn't clear to me was why we don't have labels/annotations/warning stuff, but I guess that's because it's not an alert?

One question here - I imagine we'll want to display *something* about this in the GUI console - do we expect them to trigger off this metric for example?

**Comment by jkyros** [comment]
Yeah, the other stuff that was in there was for alerting. There are different "types" of rules, it just so happens that the other rules that are there for the MCC are for the paused pool alert and not recording rules. 

I personally like that we're doing the aggregation of the pool metrics at the "rules level" rather than emitting a separate metric from the MCC, but what do I know :smile:

File: install/0000_90_machine-config-operator_01_prometheus-rules.yaml [hunk 1]
```diff
@@ -33,6 +33,10 @@ spec:
...
             description: "Machine config pools have a 'pause' feature, which allows config to be rendered, but prevents it from being rolled out to the nodes. This alert indicates that a certificate rotation has taken place, and the new kubelet-ca certificate bundle has been rendered into a machine config, but because the pool '{{$labels.pool}}' is paused, the config cannot be rolled out to the nodes in that pool. You will notice almost immediately that for nodes in pool '{{$labels.pool}}', pod logs will not be visible in the console and interactive commands (oc log, oc exec, oc debug, oc attach) will not work. You must unpause machine config pool '{{$labels.pool}}' to let the certificates through before the kube-apiserver-to-kubelet-signer certificate expires. You have approximately {{ $value | humanizeDuration }} remaining before this happens and nodes in '{{$labels.pool}}' cease to function properly." 
             runbook_url: https://github.com/openshift/blob/master/alerts/machine-config-operator/MachineConfigControllerPausedPoolKubeletCA.md
+    - name: os-image-override.rules
```
Comment (jkyros) [inline install/0000_90_machine-config-operator_01_prometheus-rules.yaml:36]: Our "standards" for prometheus metrics/rules have changed over time -- is this `name.rules` naming for the group a convention we should be following for future rules (or just recording rules?) and/or do we need to worry about retrofitting?

Comment (cgwalters) [reply install/0000_90_machine-config-operator_01_prometheus-rules.yaml:36]: > Our "standards" for prometheus metrics/rules have changed over time.

The rules for the rules

**Comment by sinnykumari** [comment]
> I'll be honest, I don't know the Prometheus stuff very well. Something that wasn't clear to me was why we don't have labels/annotations/warning stuff, but I guess that's because it's not an alert?

+1 on what John said. Also, have found https://prometheus.io/docs/prometheus/latest/configuration/recording_rules/#recording-rules helpful to understand difference between recording and alerting rules.

> One question here - I imagine we'll want to display _something_ about this in the GUI console - do we expect them to trigger off this metric for example?

For console based on what we would want to show, we can use alerting with `severty:info` for existing os_image_url_override metric or something more meaningful.

File: install/0000_90_machine-config-operator_01_prometheus-rules.yaml (hunk 1 above)
Comment (sinnykumari) [reply install/0000_90_machine-config-operator_01_prometheus-rules.yaml:36]: I am not aware of if this is a rule or not. I picked this convention from existing rules that I came across in CMO e.g. https://github.com/openshift/cluster-monitoring-operator/blob/master/assets/cluster-monitoring-operator/prometheus-rule.yaml#L14 . I know CMO team is working on a doc, so we may get better idea of best practices once available.

**Comment by sinnykumari** [comment]
@simonpasquier @jan--f can I get a review and lgtm from one of you :) This is follow-up from https://github.com/openshift/cluster-monitoring-operator/pull/1784

**Comment by jan--f** [comment]
:+1: 
/lgtm

**Pull request merged**